consumer/consumer.go

```diff
@@ -0,0 +1,192 @@
+	return c
+}
+
+// Serve runs the consumer and listens for new messages on the given topics.
+func (c *Consumer) Serve(clientID string, addrs ...string) error {
+	c.setup()
+
+	c.config = newClusterConfig(clientID)
+	topics := c.handlers.Topics()
+
+	consumerGroup := fmt.Sprintf("%s-consumer-group", clientID)
+	var err error
+	c.consumer, err = cluster.NewConsumer(
+		addrs,
+		consumerGroup,
+		topics,
+		c.config)
+	if err != nil {
+		// Note: this kind of error comparison is weird, but
+		// it's possible because sarama defines the KError
+		// type as an int16 that supports the error interface.
+		if kerr, ok := err.(sarama.KError); ok && kerr == sarama.ErrConsumerCoordinatorNotAvailable {
+			// We'll add some additional context
+			// information.  This issue comes from the
+			// RefreshCoordinator call inside sarama, but
+			// it's not reporting the root cause, sadly.
+			//
+			// Even with this information it's rather an
+			// annoying little issue.
+			err = errors.Wrap(err, "__consumer_offsets topic doesn't yet exist, either because no client has yet requested an offset, or because this consumer group is not yet functioning at startup or after rebalancing.")
+		}
+		return errors.Wrap(err, fmt.Sprintf("failed to create a consumer for topics %+v in consumer group %q", topics, consumerGroup))
+	}
+
+	err = c.handlePartitions(c.consumer.Partitions())
+	return err
+}
+
+// Stop the consumer.
+func (c *Consumer) Stop() error {
+	close(c.quit)
+	defer c.wg.Wait()
+
+	err := c.consumer.Close()
+	return errors.Wrap(err, "failed to close the consumer properly")
+}
+
+func (c *Consumer) handlePartitions(ch <-chan cluster.PartitionConsumer) error {
+	for {
+		select {
+		case part, ok := <-ch:
+			if !ok {
+				return fmt.Errorf("partition consumer channel closed")
+			}
+
+			c.wg.Add(1)
+			go func(pc cluster.PartitionConsumer) {
+				defer c.wg.Done()
+
+				c.handleMessages(pc.Messages(), c.consumer, pc)
+			}(part)
+		case <-c.quit:
+			return nil
+		}
+	}
+}
+
+func (c *Consumer) handleMessages(ch <-chan *sarama.ConsumerMessage, offset offsetStash, max highWaterMarker) {
+	for msg := range ch {
+		var attempts int
+
+		for {
+			attempts++
+
+			m := c.convertMessage(msg)
+			// Note: The second returned value is not checked because we will never receive messages for a topic
+			// that it does not have a handler for.
+			h, _ := c.handlers.Get(msg.Topic)
+			err := h.(handler.Handler).HandleMessage(m)
+			if err == nil {
+				offset.MarkOffset(msg, "")
+				if c.metrics != nil {
+					c.metrics.Reports(*m, map[string]string{
+						"attempts":        strconv.Itoa(attempts),
+						"msgOffset":       strconv.FormatInt(msg.Offset, 10),
+						"remainingOffset": strconv.FormatInt(max.HighWaterMarkOffset()-msg.Offset, 10),
+					})
+				}
+				break
+			}
+
+			select {
+			case <-time.After(c.RetryInterval):
+			case <-c.quit:
+				return
+			}
+		}
+	}
+}
+
+func (c *Consumer) convertMessage(cm *sarama.ConsumerMessage) *message.Message {
```

Comment:
:tulip: this is the sort of thing that we might want to pull out of the `Consumer` type and make public if we want people to be able to user felice in a modular way.